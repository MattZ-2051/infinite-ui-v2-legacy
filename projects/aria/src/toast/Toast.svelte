<script lang="ts">
  import { mdiClose } from '@mdi/js';
  import Icon from '$ui/icon/Icon.svelte';
  import { ToastContainer } from '$ui/toast';
</script>

<ToastContainer
  class="absolute top-0 left-0 right-0 overflow-hidden text-xl italic font-extrabold"
  let:severity
  let:message
  let:onMessageClick
  let:closeable
  let:close
  let:showProgressBar
  let:progress
  let:classes
>
  <div class="shadow relative toast toast-{severity} {classes}">
    <div class="container flex items-center justify-between">
      <div class="toast-message" on:click={onMessageClick}>{@html message}</div>
      {#if closeable}
        <button
          type="button"
          on:click={close}
          title="Close"
          class="bg-black bg-opacity-30 text-gray-900 hover:bg-opacity-60 hover:text-white focus:outline-none focus:ring-2 focus:ring-gray-200 rounded-full p-1 inline-flex items-center justify-center"
          ><Icon path={mdiClose} /></button
        >
      {/if}
    </div>
    {#if showProgressBar}
      <div
        role="progressbar"
        aria-label="notification timer"
        class="absolute bg-black bg-opacity-40 h-1 bottom-0 left-0"
        style="width: {progress * 100}%"
      />
    {/if}
  </div>
</ToastContainer>

<style>
  .toast {
    color: #000000;
  }

  .toast-danger {
    background: linear-gradient(89.89deg, #fb5543 0%, #f24c66 100%), #fc5746;
  }
  .toast-success {
    background: linear-gradient(89.89deg, #00eb7c 0%, #11d6ec 100%);
  }
  .toast-warning {
    background: linear-gradient(45deg, #ffae12, #f0da16 98.96%);
  }
  .toast-info {
    background: #5bc0de;
  }

  .toast-message :global(a) {
    text-decoration: underline;
  }

  .toast-message :global(a):hover {
    text-decoration: none;
  }
</style>
