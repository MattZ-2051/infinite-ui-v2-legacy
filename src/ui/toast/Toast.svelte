<script lang="ts">
  import { createEventDispatcher, onMount } from 'svelte';
  import { fly } from 'svelte/transition';
  import { tweened } from 'svelte/motion';
  import { get } from 'svelte/store';
  import { mdiClose } from '@mdi/js';
  import type { ToastSeverity } from './types';
  import Icon from '$ui/icon/Icon.svelte';
  import { getThemeStyle } from './themes';

  export let toastId: string;
  export let message: string;
  export let severity: ToastSeverity = 'info';
  export let duration = 5000;
  export let closeable = true;
  export let pauseOnHover = true;
  export let showProgressBar = true;

  const MIN_VALUE = 0;
  const MAX_VALUE = 1;
  const dispatch = createEventDispatcher<{ close: string }>();
  const progress = tweened(MIN_VALUE, { duration });

  $: $progress === MAX_VALUE && dispatch('close', toastId);

  onMount(() => {
    if (duration) {
      $progress = MAX_VALUE;
    }
  });

  function close(): void {
    progress.set(MAX_VALUE, { duration: 0 });
  }

  function resumeToast(): void {
    if (duration) {
      const value = get(progress);
      const remaining = duration - duration * value;
      progress.set(MAX_VALUE, { duration: remaining });
    }
  }

  function pauseToast(): void {
    if (duration) {
      const value = get(progress);
      progress.set(value, { duration: 0 });
    }
  }
</script>

<svelte:window on:focus={resumeToast} on:blur={pauseToast} />

<div
  class="shadow relative"
  style={getThemeStyle(severity)}
  on:mouseenter={pauseOnHover ? pauseToast : undefined}
  on:mouseleave={pauseOnHover ? resumeToast : undefined}
  in:fly={{ x: 250 }}
  out:fly={$progress === MAX_VALUE ? { x: 250 } : { duration: 0 }}
  role="status"
>
  <div class="container flex items-center justify-between">
    <div class="toast-message">{@html message}</div>
    {#if closeable}
      <button
        type="button"
        on:click={close}
        title="Close"
        class="bg-black bg-opacity-30 text-gray-100 hover:bg-opacity-60 hover:text-white focus:outline-none focus:ring-2 focus:ring-inset focus:ring-gray-600 rounded-full p-1 inline-flex items-center justify-center"
        ><Icon path={mdiClose} /></button
      >
    {/if}
  </div>
  {#if showProgressBar}
    <div
      role="progressbar"
      aria-label="notification timer"
      class="absolute bg-black bg-opacity-40 h-1 bottom-0 left-0"
      style="width: {$progress * 100}%"
    />
  {/if}
</div>

<style>
  .toast-message :global(a) {
    text-decoration: underline;
  }

  .toast-message :global(a):hover {
    text-decoration: none;
  }
</style>
