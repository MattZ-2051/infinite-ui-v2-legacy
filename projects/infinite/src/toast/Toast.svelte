<script lang="ts">
  import { mdiClose } from '@mdi/js';
  import Icon from '$ui/icon/Icon.svelte';
  import { ToastContainer } from '$ui/toast';
</script>

<ToastContainer
  class="absolute top-0 left-0 right-0 overflow-hidden flex flex-col gap-1"
  let:severity
  let:message
  let:onMessageClick
  let:closeable
  let:close
>
  <div class="container py-3 pl-8 pr-3 rounded-full toast toast-{severity} mx-auto">
    <div class="flex gap-x-10 items-center justify-between">
      <div class="toast-message" on:click={onMessageClick}>{@html message}</div>
      {#if closeable}
        <button
          type="button"
          on:click={close}
          title="Close"
          class="bg-black bg-opacity-30 text-gray-900 hover:bg-opacity-60 hover:text-white focus:outline-none focus:ring-2 focus:ring-gray-200 rounded-full p-1 inline-flex items-center justify-center"
          ><Icon path={mdiClose} /></button
        >
      {/if}
    </div>
  </div>
</ToastContainer>

<style>
  .toast {
    max-width: fit-content;
    color: #000000;
  }

  .toast-danger {
    background: #f87474;
  }
  .toast-success {
    background: #ddf874;
  }
  .toast-warning {
    background: #f8c374;
  }
  .toast-info {
    background: #5bc0de;
  }

  .toast-message :global(a) {
    text-decoration: underline;
  }

  .toast-message :global(a):hover {
    text-decoration: none;
  }
</style>
