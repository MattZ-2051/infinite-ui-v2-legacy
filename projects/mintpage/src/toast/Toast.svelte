<script lang="ts">
  import { mdiWindowClose } from '@mdi/js';
  import Icon from '$ui/icon/Icon.svelte';
  import { ToastContainer } from '$ui/toast';
</script>

<ToastContainer
  class="absolute top-0 left-0 right-0 overflow-hidden flex flex-col gap-1"
  let:severity
  let:message
  let:onMessageClick
  let:closeable
  let:close
  let:classes
>
  <div class="toast toast-{severity} mx-auto {classes}">
    <div class="flex items-stretch gap-x-10 items-center justify-between">
      <div class="p-6 toast-message" on:click={onMessageClick}>{@html message}</div>
      {#if closeable}
        <button type="button" on:click={close} title="Close" class="p-4 toast-close toast-close-{severity}">
          <Icon path={mdiWindowClose} />
        </button>
      {/if}
    </div>
  </div>
</ToastContainer>

<style>
  .toast {
    border-radius: 0.5rem;
    color: #000000;
    max-width: fit-content;
  }

  .toast-close:hover {
    border-radius: 0 0.5rem 0.5rem 0;
    transition: background-color 500ms;
  }

  .toast-success {
    box-shadow: 0 8px 24px 0 #255c051f;
    background: #f0fbd1;
    color: #255c05;
  }

  .toast-success:hover {
    box-shadow: 0 8px 24px 0 #255c0533;
  }

  .toast-close-success:hover {
    background-color: #def8a4;
  }

  .toast-warning {
    box-shadow: 0 8px 24px 0 #705e011f;
    background: #fdfacc;
    color: #705e01;
  }

  .toast-warning:hover {
    box-shadow: 0 8px 24px 0 #705e0133;
  }

  .toast-close-warning:hover {
    background-color: #fcf59a;
  }

  .toast-info {
    box-shadow: 0 8px 24px 0 #002a5b1f;
    background: #c9fbfa;
    color: #002a5b;
  }

  .toast-info:hover {
    box-shadow: 0 8px 24px 0 #002a5b33;
  }

  .toast-close-info:hover {
    background-color: #95f4f8;
  }

  .toast-danger {
    box-shadow: 0 8px 24px 0 #7404111f;
    background: #fee6d0;
    color: #740411;
  }

  .toast-danger:hover {
    box-shadow: 0 8px 24px 0 #74041133;
  }

  .toast-close-danger:hover {
    background-color: #fdc7a2;
  }

  .toast-message :global(a) {
    text-decoration: underline;
  }

  .toast-message :global(a):hover {
    text-decoration: none;
  }
</style>
