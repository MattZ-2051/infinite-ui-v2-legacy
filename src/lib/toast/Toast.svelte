<script context="module" lang="ts">
  import type { ToastType } from './toast.store';
  import { styles } from '$util/styles';

  const themes = {
    danger: 'linear-gradient(89.89deg, #FB5543 0%, #F24C66 100%), #FC5746;',
    success: 'linear-gradient(89.89deg, #00EB7C 0%, #11D6EC 100%);',
    warning: 'linear-gradient(45deg,#ffae12,#f0da16 98.96%)',
    info: '#5bc0de',
  };

  function getThemeStyle(key: ToastType) {
    return styles({ background: themes[key] });
  }
</script>

<script lang="ts">
  import { fly } from 'svelte/transition';

  import Icon from '$ui/icon/Icon.svelte';
  import { mdiClose } from '@mdi/js';

  import toast from './toast.store';

  function remove(id: string) {
    toast.remove(id);
  }
</script>

<div data-style="toast-container" {...$$restProps}>
  {#each $toast as item (item.id)}
    <div class="shadow" style={getThemeStyle(item.type)} transition:fly={{ x: 250 }} role="status">
      <div class="container flex items-center justify-between">
        <div>{@html item.text}</div>
        {#if item.closeable}
          <button
            type="button"
            on:click={() => remove(item.id)}
            data-style="close"
            title="Close"
            class="rounded-full p-1 inline-flex items-center justify-center"
            ><Icon path={mdiClose} /></button
          >
        {/if}
      </div>
    </div>
  {/each}
</div>

<style>
  [data-style='toast-container'] {
    display: flex;
    flex-direction: column;
  }

  [data-style='close'] {
    @apply bg-black;
    @apply bg-opacity-30;
    @apply text-gray-100;
    @apply hover:bg-opacity-60;
    @apply hover:text-white;
    @apply focus:outline-none;
    @apply focus:ring-2;
    @apply focus:ring-inset;
    @apply focus:ring-gray-600;
  }
</style>
