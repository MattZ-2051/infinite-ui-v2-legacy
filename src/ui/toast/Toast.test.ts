import { render, RenderResult, fireEvent } from '@testing-library/svelte';
// eslint-disable-next-line unicorn/prevent-abbreviations
import type { ToastProps } from './types';
import Toast from './Toast.svelte';

describe('Toast', () => {
  const onCloseMock = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const setup = ({
    message = 'content',
    toastId = '123',
    ...rest
  }: Partial<ToastProps>): RenderResult & {
    toast: HTMLElement;
    closeButton?: HTMLElement;
    progressBar?: HTMLElement;
  } => {
    const utils = render(Toast, {
      props: {
        message,
        toastId,
        ...rest,
      },
    });

    utils.component.$on('close', onCloseMock);

    const toast = utils.getByRole('status');
    const closeButton = utils.queryByTitle('Close');
    const progressBar = utils.queryByRole('progressbar');
    return { toast, closeButton, progressBar, ...utils };
  };

  it('renders correctly', () => {
    const { container, toast, closeButton, progressBar } = setup({});

    expect(container.firstChild).toMatchSnapshot();
    expect(toast).toHaveTextContent('content');
    expect(closeButton).not.toBeNull();
    expect(progressBar).not.toBeNull();
  });

  it('dispatches close event on closeButton click', async () => {
    const { closeButton } = setup({});

    expect(onCloseMock).not.toHaveBeenCalled();

    await fireEvent.click(closeButton);
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it('hides close button based on closeable prop', () => {
    const { closeButton } = setup({ closeable: false });

    expect(closeButton).toBeNull();
  });

  it('hides progressbar based on showProgressBar prop', () => {
    const { progressBar } = setup({ showProgressBar: false });

    expect(progressBar).toBeNull();
  });
});
