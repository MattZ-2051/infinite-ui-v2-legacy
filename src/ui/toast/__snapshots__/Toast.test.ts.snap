// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders correctly 1`] = `
<div>
  <div
    class="shadow relative"
    role="status"
    style="background: #5bc0de;"
  >
    <div
      class="container flex items-center justify-between"
    >
      <div>
        content
      </div>
       
      <button
        class="bg-black bg-opacity-30 text-gray-100 hover:bg-opacity-60 hover:text-white focus:outline-none focus:ring-2 focus:ring-inset focus:ring-gray-600 rounded-full p-1 inline-flex items-center justify-center"
        title="Close"
        type="button"
      >
        <svg
          class="svelte-vk0vfl"
          style="width: 1.5rem; height: 1.5rem; fill: currentColor;"
          viewBox="0 0 24 24"
        >
          
          <path
            d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
          />
        </svg>
      </button>
    </div>
     
    <div
      aria-label="notification timer"
      class="absolute bg-black bg-opacity-40 h-1 bottom-0 left-0"
      role="progressbar"
      style="width: 0%;"
    />
  </div>
</div>
`;
